File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.ts
/**
 * Definition for Node.
 * class Node {
 *     val: number
 *     left: Node | null
 *     right: Node | null
 *     next: Node | null
 *     constructor(val?: number, left?: Node, right?: Node, next?: Node) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

/**
 * Connects each node to its next right node in the same level of a binary tree.
 * Uses level-order traversal (BFS) to process nodes level by level.
 * 
 * @param root - The root node of the binary tree
 * @returns The root of the modified tree with next pointers set
 */
function connect(root: Node | null): Node | null {
    // Handle empty tree case
    if (!root) {
        return null;
    }
  
    // Initialize queue with root node for BFS traversal
    const currentLevelQueue: Node[] = [root];
  
    // Process tree level by level
    while (currentLevelQueue.length > 0) {
        // Queue to store nodes of the next level
        const nextLevelQueue: Node[] = [];
      
        // Previous node in the current level for linking
        let previousNode: Node | null = null;
      
        // Process all nodes in the current level
        for (const currentNode of currentLevelQueue) {
            // Link previous node to current node if exists
            if (previousNode) {
                previousNode.next = currentNode;
            }
          
            // Update previous node reference
            previousNode = currentNode;
          
            // Add children to next level queue if they exist
            const { left, right } = currentNode;
            if (left) {
                nextLevelQueue.push(left);
            }
            if (right) {
                nextLevelQueue.push(right);
            }
        }
      
        // Replace current level queue with next level queue
        currentLevelQueue.splice(0, currentLevelQueue.length, ...nextLevelQueue);
    }
  
    return root;
}

File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    /**
     * Connects each node to its next right node in the same level.
     * Uses BFS (level-order traversal) to process nodes level by level.
     * 
     * @param root The root of the binary tree
     * @return The root of the modified tree with next pointers set
     */
    public Node connect(Node root) {
        // Handle empty tree case
        if (root == null) {
            return root;
        }
      
        // Initialize queue for BFS traversal
        Deque<Node> queue = new ArrayDeque<>();
        queue.offer(root);
      
        // Process tree level by level
        while (!queue.isEmpty()) {
            // Track the previous node in the current level
            Node previousNode = null;
          
            // Get the number of nodes in the current level
            int levelSize = queue.size();
          
            // Process all nodes in the current level
            for (int i = 0; i < levelSize; i++) {
                // Dequeue the current node
                Node currentNode = queue.poll();
              
                // Connect previous node to current node if previous exists
                if (previousNode != null) {
                    previousNode.next = currentNode;
                }
              
                // Update previous node for next iteration
                previousNode = currentNode;
              
                // Add left child to queue if it exists
                if (currentNode.left != null) {
                    queue.offer(currentNode.left);
                }
              
                // Add right child to queue if it exists
                if (currentNode.right != null) {
                    queue.offer(currentNode.right);
                }
            }
        }
      
        return root;
    }
}